File: ec_v26/shopjoy_v260406/_apps/android/ShopjoyBoApp/app/src/main/java/com/shopjoy/bo/util/SjLog.kt
package com.shopjoy.bo.util

import android.util.Log
import com.shopjoy.bo.BuildConfig

/**
 * ShopjoyBoApp 통합 로그 유틸 (FO 와 동일 인터페이스).
 *
 * **로그 출력 형식**: `[ENV][파일명:라인:함수명] 메시지 — 변수=값`
 *
 * **방침**: 모든 로그는 빌드 환경(Debug/Release) 에 관계없이 항상 logcat 출력.
 *
 * @see com.shopjoy.fo.util.SjLog  FO 앱의 같은 유틸 — 시그니처/동작 동일
 */
object SjLog {

    /** logcat TAG — Logcat 필터에서 "ShopjoyBo" 로 BO 앱 로그만 추출. */
    private const val TAG = "ShopjoyBo"

    /** 환경 식별자: `{buildType}/{flavor}`. */
    private val ENV: String = BuildConfig.BUILD_TYPE + "/" + BuildConfig.FLAVOR

    /**
     * 함수 진입 로그.
     * @param label  함수명 표시 override.
     * @param vars   파라미터 / 진입 시점 주요 변수.
     */
    @JvmStatic
    fun fn(label: String? = null, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        val fnName = label ?: frame.methodName
        Log.d(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:$fnName] ENTER$varStr")
    }

    /** 일반 디버그 로그. */
    @JvmStatic
    fun d(msg: String, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.d(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] $msg$varStr")
    }

    /** 정보 로그 — 중요 이벤트. */
    @JvmStatic
    fun i(msg: String, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.i(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] $msg$varStr")
    }

    /** 경고 로그. */
    @JvmStatic
    fun w(msg: String, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.w(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] $msg$varStr")
    }

    /**
     * 에러 로그.
     * @param t  예외 객체 (stacktrace 가 logcat 에 자동 출력).
     */
    @JvmStatic
    fun e(msg: String, t: Throwable? = null, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.e(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] $msg$varStr", t)
    }

    /** WebView URL 추적 — `WEBVIEW.` prefix grep 가능. */
    @JvmStatic
    fun webview(action: String, url: String?, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.d(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] WEBVIEW.$action url=$url$varStr")
    }

    /** 라이프사이클 이벤트. */
    @JvmStatic
    fun lifecycle(event: String, vararg vars: Pair<String, Any?>) {
        val frame = caller(2)
        val varStr = formatVars(vars)
        Log.i(TAG, "[$ENV][${frame.fileName}:${frame.lineNumber}:${frame.methodName}] LIFECYCLE.$event$varStr")
    }

    /** 앱 시작 정보. */
    @JvmStatic
    fun appStart(buildInfo: Map<String, Any?>) {
        Log.i(TAG, "════════════════════════════════════════════════════════════════")
        Log.i(TAG, "[$ENV] APP START — ShopjoyBoApp")
        buildInfo.forEach { (k, v) -> Log.i(TAG, "[$ENV]   $k = $v") }
        Log.i(TAG, "════════════════════════════════════════════════════════════════")
    }

    /** 앱 종료 정보. */
    @JvmStatic
    fun appEnd(reason: String) {
        Log.i(TAG, "[$ENV] APP END — reason=$reason")
        Log.i(TAG, "════════════════════════════════════════════════════════════════")
    }

    /** 호출자 stacktrace frame 추출. */
    private fun caller(depth: Int): StackTraceElement {
        val trace = Thread.currentThread().stackTrace
        return trace.getOrNull(depth + 1) ?: trace.last()
    }

    /** vars 배열 포매팅. */
    private fun formatVars(vars: Array<out Pair<String, Any?>>): String {
        if (vars.isEmpty()) return ""
        val sb = StringBuilder(" — ")
        vars.forEachIndexed { i, (k, v) ->
            if (i > 0) sb.append(", ")
            sb.append(k).append("=").append(safeStr(v))
        }
        return sb.toString()
    }

    /** 200자 초과 값 절단. */
    private fun safeStr(v: Any?): String {
        if (v == null) return "null"
        val s = v.toString()
        return if (s.length > 200) s.take(200) + "...(truncated)" else s
    }
}

File: ec_v26/shopjoy_v260406/_apps_app/capacitor/ShopjoyFoApp/www/sjLog.js
// ShopjoyFoApp 통합 로그 유틸 (브라우저 WebView 환경, 순수 JavaScript)
//
// 형식: [ENV][파일명:라인:함수명] 메시지 — 변수=값
//
// ⚠️ 모든 로그는 환경(local/dev/prod) 관계없이 항상 콘솔 출력 (사용자 요구사항).
//
// 로드 순서: env.js → sjLog.js → 기타 앱 코드
//   - env.js 가 window.__APP_ENV__ / window.__APP_ENV_NAME__ 을 먼저 정의해야 함
//   - sjLog.js 는 그 값을 읽어 prefix 구성
//
// 사용 예:
//   sjLog.fn('loadPage', { url, isCold });        // 함수 진입 + 파라미터
//   sjLog.d('결제 시작', { orderId, amount });     // 일반 디버그
//   sjLog.webview('loadUrl', url);                 // WebView URL 추적 강조
//   sjLog.lifecycle('appReady');                   // 라이프사이클
//   sjLog.appStart({ extra: 'value' });            // 앱 시작 (빌드 정보 출력)
(function (global) {
  'use strict';

  // ───── 환경 식별자 ─────
  // env.js (set-env.js 가 자동 생성) 가 window.__APP_ENV_NAME__ 설정.
  // 예: 'local', 'dev', 'prod'.
  var ENV_NAME = (global.__APP_ENV_NAME__ || 'unknown');

  /**
   * Error.stack 에서 호출자 frame (file:line:fn) 추출.
   *
   * Chrome/V8 stack: "    at funcName (http://.../foo.js:42:10)"
   * Safari   stack: "funcName@http://.../foo.js:42:10"
   *
   * @param {number} depth - sjLog 메서드 자체의 깊이 (보통 2).
   * @returns {{file: string, line: string, fn: string}}
   */
  function caller(depth) {
    var err = new Error();                         // 호출 시점 stacktrace 캡처.
    var lines = (err.stack || '').split('\n');
    // 0: "Error", 1: caller(), 2: 공개 메서드, 3: 실제 호출자.
    var line = lines[depth + 2] || lines[lines.length - 1] || '';
    var match = line.match(/(?:at\s+)?(?:(\S+?)\s*\()?([^()\s]+):(\d+):(\d+)\)?$/);
    if (!match) return { file: '?', line: '?', fn: '?' };
    var url = match[2] || '';
    // URL 의 마지막 path 컴포넌트만 (= 파일명), query string 제거.
    var file = url.split(/[\/\\]/).pop().split('?')[0] || '?';
    return {
      file: file,
      line: match[3] || '?',
      // "Object.funcName" → "funcName" 정리.
      fn: (match[1] || 'anonymous').replace(/^Object\./, '')
    };
  }

  /**
   * 값을 안전한 문자열로 변환 — 200자 초과 절단, 객체는 JSON.stringify.
   * @param {*} v
   * @returns {string}
   */
  function safeStr(v) {
    if (v === null) return 'null';
    if (v === undefined) return 'undefined';
    var s;
    try {
      if (typeof v === 'string') s = v;
      else if (typeof v === 'object') s = JSON.stringify(v);  // 순환 참조 시 throw.
      else s = String(v);
    } catch (e) {
      s = '[unstringifiable: ' + (e && e.message) + ']';
    }
    return s.length > 200 ? s.slice(0, 200) + '...(truncated)' : s;
  }

  /**
   * vars 객체 → ` — k1=v1, k2=v2` 형식.
   * @param {Object<string, *>} [vars]
   * @returns {string}
   */
  function formatVars(vars) {
    if (!vars) return '';
    var keys = Object.keys(vars);
    if (keys.length === 0) return '';
    return ' — ' + keys.map(function (k) { return k + '=' + safeStr(vars[k]); }).join(', ');
  }

  /**
   * 모든 로그 메서드의 prefix: `[ENV][파일명:라인:함수명]`.
   * @param {string} [label]  함수명 override.
   */
  function formatPrefix(label) {
    var c = caller(2);
    var fn = label || c.fn;
    return '[' + ENV_NAME + '][' + c.file + ':' + c.line + ':' + fn + ']';
  }

  // ───── 공개 API ─────
  var sjLog = {

    /** 함수 진입 로그. label 로 함수명 override 가능. */
    fn: function (label, vars) {
      console.log(formatPrefix(label) + ' ENTER' + formatVars(vars));
    },

    /** 일반 디버그 로그. */
    d: function (msg, vars) {
      console.log(formatPrefix() + ' ' + msg + formatVars(vars));
    },

    /** 정보 로그 (console.info). */
    i: function (msg, vars) {
      console.info(formatPrefix() + ' ' + msg + formatVars(vars));
    },

    /** 경고 로그 (console.warn). */
    w: function (msg, vars) {
      console.warn(formatPrefix() + ' ⚠️ ' + msg + formatVars(vars));
    },

    /**
     * 에러 로그 (console.error) — Error 객체도 함께 전달하여 DevTools 가 stacktrace 펼쳐줌.
     * @param {string} msg
     * @param {Error|*} [err]
     * @param {Object} [vars]
     */
    e: function (msg, err, vars) {
      var errStr = err ? ' — error=' + safeStr(err && err.message ? err.message : err) : '';
      console.error(formatPrefix() + ' ❌ ' + msg + errStr + formatVars(vars), err || '');
    },

    /**
     * WebView URL 추적 — `WEBVIEW.` prefix 로 grep 용이.
     * @param {string} action  동작명 (예: 'loadUrl', 'navigation').
     * @param {string|null} url
     * @param {Object} [vars]
     */
    webview: function (action, url, vars) {
      console.log(formatPrefix() + ' WEBVIEW.' + action + ' url=' + safeStr(url) + formatVars(vars));
    },

    /** 라이프사이클 이벤트 — `LIFECYCLE.` prefix. */
    lifecycle: function (event, vars) {
      console.log(formatPrefix() + ' LIFECYCLE.' + event + formatVars(vars));
    },

    /**
     * 앱 시작 정보 — DevTools Console 에 빌드/환경/디바이스 정보 명시 출력.
     * @param {Object} [extra]  추가 출력할 키-값.
     */
    appStart: function (extra) {
      console.info('════════════════════════════════════════════════════════════════');
      console.info('[' + ENV_NAME + '] APP START — ShopjoyFoApp');
      // 자동 수집 정보 + extra.
      var info = Object.assign({
        APP_ENV: ENV_NAME,
        APP_NAME: (global.__APP_ENV__ && global.__APP_ENV__.APP_NAME) || '',
        BASE_URL: (global.__APP_ENV__ && global.__APP_ENV__.BASE_URL) || '',
        userAgent: (global.navigator && global.navigator.userAgent) || '',
        href: (global.location && global.location.href) || ''
      }, extra || {});
      Object.keys(info).forEach(function (k) {
        console.info('[' + ENV_NAME + ']   ' + k + ' = ' + safeStr(info[k]));
      });
      console.info('════════════════════════════════════════════════════════════════');
    },

    /** 앱 종료 정보 (beforeunload 등에서). */
    appEnd: function (reason) {
      console.info('[' + ENV_NAME + '] APP END — reason=' + safeStr(reason));
      console.info('════════════════════════════════════════════════════════════════');
    },

    /** 환경 이름 반환. */
    env: function () {
      return ENV_NAME;
    }
  };

  // 전역 노출 — WebView 안에서 window.sjLog 로 접근.
  global.sjLog = sjLog;

  // ───── 자동 에러 캐치 ─────
  // unhandled JS error / promise rejection 자동으로 sjLog.e 로 출력.
  if (global.addEventListener) {
    global.addEventListener('error', function (e) {
      sjLog.e('unhandled error', e.error || e.message, {
        filename: e.filename, lineno: e.lineno, colno: e.colno
      });
    });
    global.addEventListener('unhandledrejection', function (e) {
      sjLog.e('unhandled promise rejection', e.reason);
    });
  }

  // 로드 확인용 — DevTools Console 즉시 표시.
  console.info('[' + ENV_NAME + '] sjLog ready');
})(typeof window !== 'undefined' ? window : this);
